utils/autotailor

```diff
@@ -26,37 +27,121 @@ import xml.etree.ElementTree as ET
 import xml.dom.minidom
 
 
-ALL_NS = {
-    "xccdf-1.1": "http://checklists.nist.gov/xccdf/1.1",
-    "xccdf-1.2": "http://checklists.nist.gov/xccdf/1.2",
-}
+NS = "http://checklists.nist.gov/xccdf/1.2"
 DEFAULT_PROFILE_SUFFIX = "_customized"
+DEFAULT_REVERSE_DNS = "org.ssgproject.content"
+ROLES = ["full", "unscored", "unchecked"]
+SEVERITIES = ["unknown", "info", "low", "medium", "high"]
+
+
+def quote(string):
+    return "\"" + string + "\""
 
 
 def is_valid_xccdf_id(string):
-    return re.match(r"^xccdf_", string) is not None
+    return re.match(
+        r"^xccdf_[a-zA-Z0-9.-]+_(benchmark|profile|rule|group|value|"
+        r"testresult|tailoring)_.+",
+        string) is not None
 
 
 class Tailoring:
-    ID_NAMESPACE = "org.ssgproject.content"
-
     def __init__(self):
-        self.id_suffix = "auto_tailoring_default"
+        self.id = "xccdf_auto_tailoring_default"
+        self.reverse_dns = DEFAULT_REVERSE_DNS
         self.version = 1
-        self.profile_name = ""
+        self._profile_id = None
         self.extends = ""
         self.original_ds_filename = ""
         self.profile_title = ""
 
         self.value_changes = []
         self.rules_to_select = []
         self.rules_to_unselect = []
+        self._rule_refinements = collections.defaultdict(dict)
+
+    @property
+    def profile_id(self):
+        if self._profile_id is not None:
+            return self._profile_id
+        else:
+            return self.extends + "_customized"
+
+    @profile_id.setter
+    def profile_id(self, new_profile_id):
+        self._profile_id = new_profile_id
+
+    def rule_refinements(self, rule_id, attribute):
+        return self._rule_refinements[rule_id][attribute]
+
+    @staticmethod
+    def _find_enumeration(attribute):
+        if attribute == "role":
+            enumeration = ROLES
+        elif attribute == "severity":
+            enumeration = SEVERITIES
+        else:
+            msg = (
+                "Unsupported refine-rule attribute {attribute}".format(
+                    attribute=attribute))
+            raise ValueError(msg)
+        return enumeration
+
+    @staticmethod
+    def _validate_rule_refinement_params(rule_id, attribute, value):
+        if not is_valid_xccdf_id(rule_id):
+            msg = "Rule id '{rule_id}' is invalid!".format(rule_id=rule_id)
+            raise ValueError(msg)
+        enumeration = Tailoring._find_enumeration(attribute)
+        if value in enumeration:
+            return
+        allowed = ", ".join(map(quote, enumeration))
+        msg = (
+            "Can't refine {attribute} of rule '{rule_id}' to '{value}'. "
+            "Allowed {attribute} values are: {allowed}.".format(
+                attribute=attribute, rule_id=rule_id, value=value,
+                allowed=allowed))
```

Comment:
And this: `f"Can't refine {attribute} of rule '{rule_id}' to '{value}'. Allowed {attribute} values are: {allowed}."`.